nixos/modules/services/backup/postgresql-wal-receiver.nix

```diff
@@ -0,0 +1,210 @@
+          }
+        '';
+        description = ''
+          Environment variables passed to the service.
+          Usable parameters are listed in <link xlink:href="https://www.postgresql.org/docs/current/libpq-envars.html">Section 34.14 of the PostgreSQL manual</link>.
+        '';
+      };
+    };
+  };
+
+in {
+  options = {
+    services.postgresqlWalReceiver = {
+      receivers = mkOption {
+        type = with types; attrsOf (submodule receiverSubmodule);
+        default = { };
+        example = literalExample ''
+          {
+            main = {
+              postgresqlPackage = pkgs.postgresql_9_6;
+              directory = /mnt/pg_wal/main/;
+              slot = "main_wal_receiver";
+              connection = "postgresql://user@somehost";
+            };
+          }
+        '';
+        description = ''
+          PostgreSQL WAL receivers.
+          Stream write-ahead logs from a PostgreSQL server using <command>pg_receivewal</command> (formerly <command>pg_receivexlog</command>).
+          See <link xlink:href="https://www.postgresql.org/docs/current/app-pgreceivewal.html">the man page</link> for more information.
+        '';
+      };
+    };
+  };
+
+  config = let
+    receivers = config.services.postgresqlWalReceiver.receivers;
+  in mkIf (receivers != { }) {
+    users = {
+      groups.postgresql-wal-receiver = { };
+      users.postgresql-wal-receiver = {
+        description = "PostgreSQL WAL receiver";
+        group = "postgresql-wal-receiver";
+        isSystemUser = true;
+      };
+    };
+
+    assertions = concatLists (attrsets.mapAttrsToList (name: config: [
+      {
+        assertion = !(config.noSync && config.synchronous);
+        message = "noSync is incompatible with synchronous.";
+      }
+      {
+        assertion = config.noSync -> versionAtLeast config.postgresqlPackage.version "11";
+        message = "noSync requires PostgreSQL version >= 11.";
+      }
+      {
+        assertion = config.compress > 0 -> versionAtLeast config.postgresqlPackage.version "10";
+        message = "compress requires PostgreSQL version >= 10.";
+      }
+    ]) receivers);
+
+    systemd.services = with attrsets; mapAttrs' (name: config: nameValuePair "postgresql-wal-receiver-${name}" {
+      description = "PostgreSQL WAL receiver (${name})";
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        User = user;
+        Group = group;
+        KillSignal = "SIGINT";
+        Restart = "always";
+        RestartSec = 5;
+      };
+
+      inherit (config) environment;
+
+      preStart = ''
+        mkdir -p ${escapeShellArg config.directory}
+      '';
+
+      script = let
+        receiverCommand = postgresqlPackage:
+         if (versionAtLeast postgresqlPackage.version "10")
+           then "${postgresqlPackage}/bin/pg_receivewal"
+           else "${postgresqlPackage}/bin/pg_receivexlog";
+      in ''
+        ${receiverCommand config.postgresqlPackage} \
+          --no-password \
```

Comment:
maybe a generic `extraArgs` would be great here, maybe some new flags will be added in future to this command.

Maybe similar to https://github.com/NixOS/nixpkgs/pull/65245/commits/7e4e37fff4cd97cc7fed26c9e1060761ed1379a7, maybe more sophisticated.

Reply:
Added.